File: Lab12/app/src/main/java/com/example/lab12/MainActivity.java
package com.example.lab12;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.os.Bundle;
import android.util.Log;
import com.google.gson.Gson;

import java.io.IOException;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;


public class MainActivity extends AppCompatActivity {

    class Data{
        Result result;

        class Result{
            Results[] results;

            class Results{
                String Station;
                String Destination;
            }
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        findViewById(R.id.btn_query).setOnClickListener(view -> {
            String URL="https://lab12-api.web.app/";

            Request request =new Request.Builder().url(URL).build();
            new OkHttpClient().newCall(request).enqueue(new Callback(){
               @Override
               public void onResponse(@NonNull Call call, @NonNull Response response)throws IOException{
                   if(response.code()==200){
                       if(response.body()==null) return;
                       Data data = new Gson().fromJson(response.body().string(),Data.class);
                       final String[] items =new String[data.result.results.length];

                       for(int i=0;i<items.length;i++){
                           items[i]="\n列車機將進入:"+data.result.results[i].Station+"\n列車行止目的地:"+data.result.results[i].Destination;
                       }

                       runOnUiThread(()->{
                           new AlertDialog.Builder(MainActivity.this).setTitle("台北捷運列車到站站名").setItems(items,null).show();
                       });
                   }else if(!response.isSuccessful()){
                       Log.e("伺服器錯誤",response.code()+" "+response.message());
                   }else{
                       Log.e("其他錯誤",response.code()+" "+response.message());
                   }
               }
               @Override
                public void onFailure(@NonNull Call call,@NonNull IOException e){
                   Log.e("查詢失敗",e.getMessage());
               }
            });
        });
    }
}

File: Lab17_withKotlin/app/src/main/java/com/stella/lab17_withkotlin/MainActivity.kt
package com.stella.lab17_withkotlin

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import com.google.gson.Gson
import okhttp3.*
import java.io.IOException

class MainActivity : AppCompatActivity() {
    private lateinit var btn_query: Button

    class MyObject {
        lateinit var result: Result
        class Result {
            lateinit var records: Array<Record>
            class Record {
                var SiteName = ""
                var Status = ""
            }
        }
    }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        btn_query = findViewById(R.id.btn_query)
        btn_query.setOnClickListener {

            btn_query.isEnabled = false

            sendRequest()
        }
    }

    private fun sendRequest() {

        val url = "https://api.italkutalk.com/api/air"

        val req = Request.Builder().url(url).build()

        OkHttpClient().newCall(req).enqueue(object : Callback {

            override fun onResponse(call: Call, response: Response) {

                val json = response.body?.string()
                val myObject = Gson().fromJson(json, MyObject::class.java)

                showDialog(myObject)
            }

            override fun onFailure(call: Call, e: IOException) {
                runOnUiThread {

                    btn_query.isEnabled = true
                    Toast.makeText(this@MainActivity,
                        "查詢失敗$e", Toast.LENGTH_SHORT).show()
                }
            }
        })
    }

    private fun showDialog(myObject: MyObject) {

        val items = arrayOfNulls<String>(myObject.result.records.size)

        myObject.result.records.forEachIndexed { index, data ->
            items[index] = "地區：${data.SiteName}, 狀態：${data.Status}"
        }

        runOnUiThread {

            btn_query.isEnabled = true

            AlertDialog.Builder(this@MainActivity)
                .setTitle("臺北市空氣品質")
                .setItems(items, null)
                .show()
        }
    }
}